File: Models/AppTheme.swift
import Foundation

struct ThemePalette: Codable, Equatable {
    var backgroundTop: ThemeColor
    var backgroundBottom: ThemeColor
    var surfacePrimary: ThemeColor
    var surfaceSecondary: ThemeColor
    var card: ThemeColor
    var cardCurrent: ThemeColor
    var accent: ThemeColor
    var danger: ThemeColor
    var textPrimary: ThemeColor
    var textSecondary: ThemeColor

    private enum CodingKeys: String, CodingKey {
        case backgroundTop
        case backgroundBottom
        case surfacePrimary
        case surfaceSecondary
        case card
        case cardCurrent
        case accent
        case danger
        case textPrimary
        case textSecondary
    }

    init(
        backgroundTop: ThemeColor,
        backgroundBottom: ThemeColor,
        surfacePrimary: ThemeColor,
        surfaceSecondary: ThemeColor,
        card: ThemeColor,
        cardCurrent: ThemeColor,
        accent: ThemeColor,
        danger: ThemeColor,
        textPrimary: ThemeColor,
        textSecondary: ThemeColor
    ) {
        self.backgroundTop = backgroundTop
        self.backgroundBottom = backgroundBottom
        self.surfacePrimary = surfacePrimary
        self.surfaceSecondary = surfaceSecondary
        self.card = card
        self.cardCurrent = cardCurrent
        self.accent = accent
        self.danger = danger
        self.textPrimary = textPrimary
        self.textSecondary = textSecondary
    }

    init(from decoder: Decoder) throws {
        let defaults = ThemeRenderer.defaultTheme.palette
        let container = try decoder.container(keyedBy: CodingKeys.self)
        backgroundTop = try container.decodeIfPresent(ThemeColor.self, forKey: .backgroundTop) ?? defaults.backgroundTop
        backgroundBottom = try container.decodeIfPresent(ThemeColor.self, forKey: .backgroundBottom) ?? defaults.backgroundBottom
        surfacePrimary = try container.decodeIfPresent(ThemeColor.self, forKey: .surfacePrimary) ?? defaults.surfacePrimary
        surfaceSecondary = try container.decodeIfPresent(ThemeColor.self, forKey: .surfaceSecondary) ?? defaults.surfaceSecondary
        card = try container.decodeIfPresent(ThemeColor.self, forKey: .card) ?? defaults.card
        cardCurrent = try container.decodeIfPresent(ThemeColor.self, forKey: .cardCurrent) ?? defaults.cardCurrent
        accent = try container.decodeIfPresent(ThemeColor.self, forKey: .accent) ?? defaults.accent
        danger = try container.decodeIfPresent(ThemeColor.self, forKey: .danger) ?? defaults.danger
        textPrimary = try container.decodeIfPresent(ThemeColor.self, forKey: .textPrimary) ?? defaults.textPrimary
        textSecondary = try container.decodeIfPresent(ThemeColor.self, forKey: .textSecondary) ?? defaults.textSecondary
    }
}

enum BackgroundMode: String, CaseIterable, Codable {
    case none
    case image
}

enum BackgroundLayoutMode: String, CaseIterable, Codable, Identifiable {
    case fill
    case fit
    case center
    case tile

    var id: String { rawValue }
}

struct BackgroundConfig: Codable, Equatable {
    var mode: BackgroundMode
    var imageBookmarkData: Data?
    var imageOriginalPath: String?
    var layoutMode: BackgroundLayoutMode
    var opacity: Double
    var dimOverlay: Double
    var blurRadius: Double

    private enum CodingKeys: String, CodingKey {
        case mode
        case imageBookmarkData
        case imageOriginalPath
        case layoutMode
        case opacity
        case dimOverlay
        case blurRadius
    }

    init(
        mode: BackgroundMode,
        imageBookmarkData: Data?,
        imageOriginalPath: String?,
        layoutMode: BackgroundLayoutMode,
        opacity: Double,
        dimOverlay: Double,
        blurRadius: Double
    ) {
        self.mode = mode
        self.imageBookmarkData = imageBookmarkData
        self.imageOriginalPath = imageOriginalPath
        self.layoutMode = layoutMode
        self.opacity = opacity
        self.dimOverlay = dimOverlay
        self.blurRadius = blurRadius
    }

    init(from decoder: Decoder) throws {
        let defaults = ThemeRenderer.defaultTheme.background
        let container = try decoder.container(keyedBy: CodingKeys.self)
        mode = try container.decodeIfPresent(BackgroundMode.self, forKey: .mode) ?? defaults.mode
        imageBookmarkData = try container.decodeIfPresent(Data.self, forKey: .imageBookmarkData)
        imageOriginalPath = try container.decodeIfPresent(String.self, forKey: .imageOriginalPath)
        layoutMode = try container.decodeIfPresent(BackgroundLayoutMode.self, forKey: .layoutMode) ?? defaults.layoutMode
        opacity = try container.decodeIfPresent(Double.self, forKey: .opacity) ?? defaults.opacity
        dimOverlay = try container.decodeIfPresent(Double.self, forKey: .dimOverlay) ?? defaults.dimOverlay
        blurRadius = try container.decodeIfPresent(Double.self, forKey: .blurRadius) ?? defaults.blurRadius
    }
}

enum InterfaceDensity: String, CaseIterable, Codable, Identifiable {
    case compact
    case normal
    case spacious

    var id: String { rawValue }
}

enum PanelBlurStrength: String, CaseIterable, Codable, Identifiable {
    case low
    case medium
    case high

    var id: String { rawValue }
}

struct UIChromeConfig: Codable, Equatable {
    var accentIntensity: Double
    var panelOpacity: Double
    var panelBlurStrength: PanelBlurStrength
    var cornerRadius: Double
    var density: InterfaceDensity

    private enum CodingKeys: String, CodingKey {
        case accentIntensity
        case panelOpacity
        case panelBlurStrength
        case cornerRadius
        case density
    }

    init(
        accentIntensity: Double,
        panelOpacity: Double,
        panelBlurStrength: PanelBlurStrength,
        cornerRadius: Double,
        density: InterfaceDensity
    ) {
        self.accentIntensity = accentIntensity
        self.panelOpacity = panelOpacity
        self.panelBlurStrength = panelBlurStrength
        self.cornerRadius = cornerRadius
        self.density = density
    }

    init(from decoder: Decoder) throws {
        let defaults = ThemeRenderer.defaultTheme.chrome
        let container = try decoder.container(keyedBy: CodingKeys.self)
        accentIntensity = try container.decodeIfPresent(Double.self, forKey: .accentIntensity) ?? defaults.accentIntensity
        panelOpacity = try container.decodeIfPresent(Double.self, forKey: .panelOpacity) ?? defaults.panelOpacity
        panelBlurStrength = try container.decodeIfPresent(PanelBlurStrength.self, forKey: .panelBlurStrength) ?? defaults.panelBlurStrength
        cornerRadius = try container.decodeIfPresent(Double.self, forKey: .cornerRadius) ?? defaults.cornerRadius
        density = try container.decodeIfPresent(InterfaceDensity.self, forKey: .density) ?? defaults.density
    }
}

enum ThemePreset: String, CaseIterable, Codable, Identifiable {
    case classicDungeon
    case tavernEmber
    case moonlitCrypt
    case forestMist

    var id: String { rawValue }
}

struct CustomThemePreset: Codable, Equatable, Identifiable {
    var id: UUID
    var name: String
    var theme: AppTheme

    init(id: UUID = UUID(), name: String, theme: AppTheme) {
        self.id = id
        self.name = name
        self.theme = theme
    }
}

struct AppTheme: Codable, Equatable {
    var preset: ThemePreset?
    var palette: ThemePalette
    var background: BackgroundConfig
    var chrome: UIChromeConfig

    private enum CodingKeys: String, CodingKey {
        case preset
        case palette
        case background
        case chrome
    }

    init(
        preset: ThemePreset?,
        palette: ThemePalette,
        background: BackgroundConfig,
        chrome: UIChromeConfig
    ) {
        self.preset = preset
        self.palette = palette
        self.background = background
        self.chrome = chrome
    }

    init(from decoder: Decoder) throws {
        let defaults = ThemeRenderer.defaultTheme
        let container = try decoder.container(keyedBy: CodingKeys.self)
        preset = try container.decodeIfPresent(ThemePreset.self, forKey: .preset) ?? defaults.preset
        palette = try container.decodeIfPresent(ThemePalette.self, forKey: .palette) ?? defaults.palette
        background = try container.decodeIfPresent(BackgroundConfig.self, forKey: .background) ?? defaults.background
        chrome = try container.decodeIfPresent(UIChromeConfig.self, forKey: .chrome) ?? defaults.chrome
    }
}

File: windows/src/DungeonSoundboard.Core/Models/AppTheme.cs
using System.Text.Json.Serialization;

namespace DungeonSoundboard.Core.Models;

public sealed class ThemePalette : IEquatable<ThemePalette>
{
    public ThemeColor BackgroundTop { get; set; } = new(0.102, 0.078, 0.071);
    public ThemeColor BackgroundBottom { get; set; } = new(0.161, 0.122, 0.102);
    public ThemeColor SurfacePrimary { get; set; } = new(0.180, 0.141, 0.122);
    public ThemeColor SurfaceSecondary { get; set; } = new(0.220, 0.169, 0.141);
    public ThemeColor Card { get; set; } = new(0.239, 0.188, 0.161);
    public ThemeColor CardCurrent { get; set; } = new(0.322, 0.239, 0.169);
    public ThemeColor Accent { get; set; } = new(0.820, 0.671, 0.322);
    public ThemeColor Danger { get; set; } = new(0.722, 0.278, 0.239);
    public ThemeColor TextPrimary { get; set; } = new(0.949, 0.910, 0.839);
    public ThemeColor TextSecondary { get; set; } = new(0.741, 0.690, 0.620);

    public bool Equals(ThemePalette? other)
    {
        return other is not null
            && BackgroundTop.Equals(other.BackgroundTop)
            && BackgroundBottom.Equals(other.BackgroundBottom)
            && SurfacePrimary.Equals(other.SurfacePrimary)
            && SurfaceSecondary.Equals(other.SurfaceSecondary)
            && Card.Equals(other.Card)
            && CardCurrent.Equals(other.CardCurrent)
            && Accent.Equals(other.Accent)
            && Danger.Equals(other.Danger)
            && TextPrimary.Equals(other.TextPrimary)
            && TextSecondary.Equals(other.TextSecondary);
    }

    public override bool Equals(object? obj) => Equals(obj as ThemePalette);

    public override int GetHashCode() => HashCode.Combine(BackgroundTop, BackgroundBottom, SurfacePrimary, Accent, TextPrimary);
}

[JsonConverter(typeof(JsonStringEnumConverter<BackgroundMode>))]
public enum BackgroundMode
{
    [JsonStringEnumMemberName("none")]
    None,

    [JsonStringEnumMemberName("image")]
    Image
}

[JsonConverter(typeof(JsonStringEnumConverter<BackgroundLayoutMode>))]
public enum BackgroundLayoutMode
{
    [JsonStringEnumMemberName("fill")]
    Fill,

    [JsonStringEnumMemberName("fit")]
    Fit,

    [JsonStringEnumMemberName("center")]
    Center,

    [JsonStringEnumMemberName("tile")]
    Tile
}

public sealed class BackgroundConfig : IEquatable<BackgroundConfig>
{
    public BackgroundMode Mode { get; set; } = BackgroundMode.None;
    public byte[]? ImageBookmarkData { get; set; }
    public string? ImageOriginalPath { get; set; }
    public BackgroundLayoutMode LayoutMode { get; set; } = BackgroundLayoutMode.Fill;
    public double Opacity { get; set; } = 0.72;
    public double DimOverlay { get; set; } = 0.30;
    public double BlurRadius { get; set; }

    public bool Equals(BackgroundConfig? other)
    {
        if (other is null)
        {
            return false;
        }

        return Mode == other.Mode
            && BytesEqual(ImageBookmarkData, other.ImageBookmarkData)
            && ImageOriginalPath == other.ImageOriginalPath
            && LayoutMode == other.LayoutMode
            && Opacity.Equals(other.Opacity)
            && DimOverlay.Equals(other.DimOverlay)
            && BlurRadius.Equals(other.BlurRadius);
    }

    public override bool Equals(object? obj) => Equals(obj as BackgroundConfig);

    public override int GetHashCode() => HashCode.Combine(Mode, ImageOriginalPath, LayoutMode, Opacity, DimOverlay, BlurRadius);

    private static bool BytesEqual(byte[]? left, byte[]? right)
    {
        if (left is null || right is null)
        {
            return left is null && right is null;
        }

        return left.SequenceEqual(right);
    }
}

[JsonConverter(typeof(JsonStringEnumConverter<InterfaceDensity>))]
public enum InterfaceDensity
{
    [JsonStringEnumMemberName("compact")]
    Compact,

    [JsonStringEnumMemberName("normal")]
    Normal,

    [JsonStringEnumMemberName("spacious")]
    Spacious
}

[JsonConverter(typeof(JsonStringEnumConverter<PanelBlurStrength>))]
public enum PanelBlurStrength
{
    [JsonStringEnumMemberName("low")]
    Low,

    [JsonStringEnumMemberName("medium")]
    Medium,

    [JsonStringEnumMemberName("high")]
    High
}

public sealed class UIChromeConfig : IEquatable<UIChromeConfig>
{
    public double AccentIntensity { get; set; } = 0.60;
    public double PanelOpacity { get; set; } = 0.86;
    public PanelBlurStrength PanelBlurStrength { get; set; } = PanelBlurStrength.Medium;
    public double CornerRadius { get; set; } = 16;
    public InterfaceDensity Density { get; set; } = InterfaceDensity.Normal;

    public bool Equals(UIChromeConfig? other)
    {
        return other is not null
            && AccentIntensity.Equals(other.AccentIntensity)
            && PanelOpacity.Equals(other.PanelOpacity)
            && PanelBlurStrength == other.PanelBlurStrength
            && CornerRadius.Equals(other.CornerRadius)
            && Density == other.Density;
    }

    public override bool Equals(object? obj) => Equals(obj as UIChromeConfig);

    public override int GetHashCode() => HashCode.Combine(AccentIntensity, PanelOpacity, PanelBlurStrength, CornerRadius, Density);
}

[JsonConverter(typeof(JsonStringEnumConverter<ThemePreset>))]
public enum ThemePreset
{
    [JsonStringEnumMemberName("classicDungeon")]
    ClassicDungeon,

    [JsonStringEnumMemberName("tavernEmber")]
    TavernEmber,

    [JsonStringEnumMemberName("moonlitCrypt")]
    MoonlitCrypt,

    [JsonStringEnumMemberName("forestMist")]
    ForestMist
}

public sealed class CustomThemePreset : IEquatable<CustomThemePreset>
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "";
    public AppTheme Theme { get; set; } = ThemeRenderer.DefaultTheme;

    public bool Equals(CustomThemePreset? other)
    {
        return other is not null && Id == other.Id && Name == other.Name && Theme.Equals(other.Theme);
    }

    public override bool Equals(object? obj) => Equals(obj as CustomThemePreset);

    public override int GetHashCode() => HashCode.Combine(Id, Name, Theme);
}

public sealed class AppTheme : IEquatable<AppTheme>
{
    public ThemePreset? Preset { get; set; } = ThemePreset.ClassicDungeon;
    public ThemePalette Palette { get; set; } = new();
    public BackgroundConfig Background { get; set; } = new();
    public UIChromeConfig Chrome { get; set; } = new();

    public AppTheme Clone()
    {
        return new AppTheme
        {
            Preset = Preset,
            Palette = new ThemePalette
            {
                BackgroundTop = new ThemeColor(Palette.BackgroundTop.Red, Palette.BackgroundTop.Green, Palette.BackgroundTop.Blue, Palette.BackgroundTop.Alpha),
                BackgroundBottom = new ThemeColor(Palette.BackgroundBottom.Red, Palette.BackgroundBottom.Green, Palette.BackgroundBottom.Blue, Palette.BackgroundBottom.Alpha),
                SurfacePrimary = new ThemeColor(Palette.SurfacePrimary.Red, Palette.SurfacePrimary.Green, Palette.SurfacePrimary.Blue, Palette.SurfacePrimary.Alpha),
                SurfaceSecondary = new ThemeColor(Palette.SurfaceSecondary.Red, Palette.SurfaceSecondary.Green, Palette.SurfaceSecondary.Blue, Palette.SurfaceSecondary.Alpha),
                Card = new ThemeColor(Palette.Card.Red, Palette.Card.Green, Palette.Card.Blue, Palette.Card.Alpha),
                CardCurrent = new ThemeColor(Palette.CardCurrent.Red, Palette.CardCurrent.Green, Palette.CardCurrent.Blue, Palette.CardCurrent.Alpha),
                Accent = new ThemeColor(Palette.Accent.Red, Palette.Accent.Green, Palette.Accent.Blue, Palette.Accent.Alpha),
                Danger = new ThemeColor(Palette.Danger.Red, Palette.Danger.Green, Palette.Danger.Blue, Palette.Danger.Alpha),
                TextPrimary = new ThemeColor(Palette.TextPrimary.Red, Palette.TextPrimary.Green, Palette.TextPrimary.Blue, Palette.TextPrimary.Alpha),
                TextSecondary = new ThemeColor(Palette.TextSecondary.Red, Palette.TextSecondary.Green, Palette.TextSecondary.Blue, Palette.TextSecondary.Alpha)
            },
            Background = new BackgroundConfig
            {
                Mode = Background.Mode,
                ImageBookmarkData = Background.ImageBookmarkData?.ToArray(),
                ImageOriginalPath = Background.ImageOriginalPath,
                LayoutMode = Background.LayoutMode,
                Opacity = Background.Opacity,
                DimOverlay = Background.DimOverlay,
                BlurRadius = Background.BlurRadius
            },
            Chrome = new UIChromeConfig
            {
                AccentIntensity = Chrome.AccentIntensity,
                PanelOpacity = Chrome.PanelOpacity,
                PanelBlurStrength = Chrome.PanelBlurStrength,
                CornerRadius = Chrome.CornerRadius,
                Density = Chrome.Density
            }
        };
    }

    public bool Equals(AppTheme? other)
    {
        return other is not null
            && Preset == other.Preset
            && Palette.Equals(other.Palette)
            && Background.Equals(other.Background)
            && Chrome.Equals(other.Chrome);
    }

    public override bool Equals(object? obj) => Equals(obj as AppTheme);

    public override int GetHashCode() => HashCode.Combine(Preset, Palette, Background, Chrome);
}
